plugins/mediatek-scaler/fu-mediatek-scaler-device.c

```diff
@@ -0,0 +1,971 @@
+		if (!fu_mediatek_scaler_device_ddc_write(self, st_req, error)) {
+			g_prefix_error(error, "failed to send firmware to device: ");
+			return FALSE;
+		}
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_get_staged_data(FuMediatekScalerDevice *self,
+					  guint16 *chksum,
+					  guint32 *pktcnt,
+					  GError **error)
+{
+	g_autoptr(GByteArray) st_req = fu_struct_ddc_cmd_new();
+	g_autoptr(GByteArray) st_res = NULL;
+
+	fu_struct_ddc_cmd_set_vcp_code(st_req, FU_DDC_VCP_CODE_GET_STAGED);
+	st_res = fu_mediatek_scaler_device_ddc_read(self, st_req, error);
+	if (st_res == NULL)
+		return FALSE;
+	if (!fu_memread_uint16_safe(st_res->data, st_res->len, 2, chksum, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	if (!fu_memread_uint32_safe(st_res->data, st_res->len, 4, pktcnt, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_check_sent_info(FuMediatekScalerDevice *self,
+					  GBytes *fwdata,
+					  GError **error)
+{
+	guint16 chksum = 0;
+	guint32 pktcnt = 0;
+
+	if (!fu_mediatek_scaler_device_get_staged_data(self, &chksum, &pktcnt, error))
+		return FALSE;
+
+	/* verify the staged packets on chip */
+	if (g_bytes_get_size(fwdata) != pktcnt) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "failed data verification, sent size: %lu, ack size: %u",
+			    g_bytes_get_size(fwdata),
+			    pktcnt);
+		return FALSE;
+	}
+
+	/* verify the checksum on chip */
+	if (fu_sum16_bytes(fwdata) != chksum) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "failed data checksum comparison, expected: %u, got: %u",
+			    fu_sum16_bytes(fwdata),
+			    chksum);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_run_isp(FuMediatekScalerDevice *self, guint16 chksum, GError **error)
+{
+	g_autoptr(GByteArray) st_req = fu_struct_ddc_cmd_new();
+	fu_struct_ddc_cmd_set_vcp_code(st_req, FU_DDC_VCP_CODE_COMMIT_FW);
+	fu_byte_array_append_uint16(st_req, chksum, G_LITTLE_ENDIAN);
+	return fu_mediatek_scaler_device_ddc_write(self, st_req, error);
+}
+
+static gboolean
+fu_mediatek_scaler_device_commit_firmware(FuMediatekScalerDevice *self, GBytes *fw, GError **error)
+{
+	if (!(fu_mediatek_scaler_device_run_isp(self, fu_sum16_bytes(fw), error))) {
+		g_prefix_error(error, "failed to commit firmware: ");
+		return FALSE;
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_mediatek_scaler_device_set_isp_reboot(FuMediatekScalerDevice *self, GError **error)
+{
+	g_autoptr(GByteArray) st_req = fu_struct_ddc_cmd_new();
+	g_autoptr(GError) error_local = NULL;
+
+	/* device will reboot after this, so the write will timed out fail */
+	fu_struct_ddc_cmd_set_vcp_code(st_req, FU_DDC_VCP_CODE_REBOOT);
+	if (!(fu_mediatek_scaler_device_ddc_write(self, st_req, &error_local))) {
+		if (!(g_error_matches(error_local, G_IO_ERROR, G_IO_ERROR_FAILED) &&
+		      errno == ETIMEDOUT)) {
```

Comment:
this seems a bit fragile, i think we should do https://github.com/fwupd/fwupd/pull/6202 instead.

Reply:
_narrator_: he did that